### DISCUSSION TITLE
setValue then trigger the onChange event?

### DISCUSSION THREAD

**Discussion by torian257x** [react-hook-form/react-hook-form#3429] [Q&A] [answered]
so my code is very spread out.

Is there a possibility that I can trigger the onChange event as soon as I call a setValue?

I want to setValue and call onChange after it set its value and **no, I don't have easy access to the onchange handler of all of my inputs that are spread out over many nested components**.

I tried this:

```
  const methods = useForm();
  const formState = methods.formState;

  const testClick = () => {
    console.error("formState:", formState);
    setValue('pi__comments__c', 'adsf asdf adsf asdf 333344', {
      shouldDirty: true}).then(function(){

      console.error("formState22:", formState.dirtyFields);
      console.error("formState22 isDirty:", formState.isDirty);
    });

  };

return <><button onClick={testClick}>buttonnn</button>
 <input name="pi__comments__c" ...>
</>
```

`formState.isDirty` is empty object as is `formState.dirtyFields`

we should have the reference to the fields somewhere, what I want is to call `setValue` and then call the `onChange` of the reference to the input field. Because I have logic like "if fieldA has value 'foo' then show fieldB" on multiple onchange handlers

I hope there is a simple solution

**Comment by bluebill1049** [comment]
<img width="740" alt="Screen Shot 2020-11-12 at 12 43 35 pm" src="https://user-images.githubusercontent.com/10513364/98884694-b0114f80-24e4-11eb-9e32-9834e10cf5af.png">

**Reply by torian257x** [reply]
@bluebill1049 I dont need field validation (well maybe I do) what I need is to run the onChange handler of the input field.

Like I have new use case where all form data is loaded from the internet. I call setValue for each field. But the fields have logic

"Show state if country == usa" 

but I do

```
setValue('country', 'usa');
setValue('state','alabama');
```
but `setValue('state','alabama');` doesnt work because it is

```
{needsState && <input name="state" type="text"/>
```

and needsState is local state in nested component.

e.g.

```
const checkIfNeedsState = (ev) =>{
const mycountry = ev.target.value;
if(mycountry == 'usa'){
setNeedsState(true);
}
}

<input name="country" onChange={checkIfNeedsState} type="text/>
{needsState && <input name="state" type="text"/>
```

this is only 1 example, I have more places with things like this where show/hide more fields depending on input

**Reply by bluebill1049** [reply]
then you should probably use valdiate function and getValue() to read other input's value and do your validation.

**Reply by torian257x** [reply]
@bluebill1049 I don't understand. You want to say remove `needsState` and make "display: none" instead and change everything?
I dont like to change so much. problem is as well, state is a select field. And has different values for say "country==canada". I need rerender.

Thus I wonder if there is access to the field refs somehow so I can trigger change?
alternative is maybe to make jquery javascript and do `$('select[name="country"]).trigger('change')`. I think very ugly, but seems the only way?

**Reply by torian257x** [reply]
but why isDirty and dirtyFields doesnt work? or am I using it wrong?

**Reply by torian257x** [reply]
`$('select[name="country"]).trigger('change')` doesnt work as react has its own onchange handler I think

**Reply by torian257x** [reply]
ah I think I understand now. Instead of  `<select onChange="checkIfNeedsState" ...` I do `<select register={ref({validate:checkIfNeedsState)} ...` for me is ok. Still worried because now I need ot change so much and is not really "validate" is more like "validateAndSetNewStateVariables" so not "great" feels a bit hacky

I still think somehow a way to access ref would be good. `getRefs` maybe, like `getValues`

**Reply by JoshuaSunsheng** [reply]
> <img alt="Screen Shot 2020-11-12 at 12 43 35 pm" width="740" src="https://user-images.githubusercontent.com/10513364/98884694-b0114f80-24e4-11eb-9e32-9834e10cf5af.png">

Wonderful solution. This is exactly what I need for a Select component. After user select one option, the js set the input value and it should make the error msg invisible.

**Comment by torian257x** [comment, marked as answer]
Instead of <select onChange="checkIfNeedsState" ... I do <select register={ref({validate:checkIfNeedsState)} ... for me is ok. Still worried because now I need ot change so much and is not really "validate" is more like "validateAndSetNewStateVariables" so not "great" feels a bit hacky

and now I have to trigger validate as well onChange or manually call it. So triggers twice (one time in validate, one time in onChange). Brrr... 

I still think somehow a way to access ref would be good. getRefs maybe, like getValues.

**Comment by shubhamkhunt04** [comment]
`setValue(name, value, { shouldValidate: true }` is work for me.
Thank you :D
